File: ios/Classes/Config/TunnelConfig.swift
import Foundation

/// Immutable shaping parameters shared with the packet tunnel provider.
public struct TunnelConfig: Codable {
  public var mode: String
  public var latencyMs: Double
  public var downloadMbps: Double
  public var uploadMbps: Double
  public var jitterMs: Double
  public var packetLoss: Double
  public var isOffline: Bool

  public static func from(map: [String: Any]?) -> TunnelConfig {
    let download = (map?["downloadMbps"] as? Double) ?? -1
    let upload = (map?["uploadMbps"] as? Double) ?? -1
    return TunnelConfig(
      mode: map?["mode"] as? String ?? "normal",
      latencyMs: map?["latencyMs"] as? Double ?? 0,
      downloadMbps: download < 0 ? .infinity : download,
      uploadMbps: upload < 0 ? .infinity : upload,
      jitterMs: map?["jitterMs"] as? Double ?? 0,
      packetLoss: map?["packetLoss"] as? Double ?? 0,
      isOffline: map?["isOffline"] as? Bool ?? false
    )
  }

  public func toProviderDict() -> [String: Any] {
    [
      "mode": mode,
      "latencyMs": latencyMs,
      "downloadMbps": downloadMbps.isFinite ? downloadMbps : -1,
      "uploadMbps": uploadMbps.isFinite ? uploadMbps : -1,
      "jitterMs": jitterMs,
      "packetLoss": packetLoss,
      "isOffline": isOffline,
    ]
  }
}

File: android/src/main/kotlin/com/darkmintis/network_simulator/config/TunnelConfig.kt
package com.darkmintis.network_simulator.config

/**
 * Immutable tunnel shaping parameters mirrored from Dart [NetworkSimulatorConfig].
 */
data class TunnelConfig(
    val mode: String = "normal",
    val latencyMs: Double = 0.0,
    val downloadMbps: Double = Double.POSITIVE_INFINITY,
    val uploadMbps: Double = Double.POSITIVE_INFINITY,
    val jitterMs: Double = 0.0,
    val packetLoss: Double = 0.0,
    val isOffline: Boolean = false,
) {
    companion object {
        fun fromMap(map: Map<*, *>?): TunnelConfig {
            if (map == null) return TunnelConfig()
            return TunnelConfig(
                mode = map["mode"] as? String ?: "normal",
                latencyMs = (map["latencyMs"] as? Number)?.toDouble() ?: 0.0,
                downloadMbps = mbpsFrom(map["downloadMbps"]),
                uploadMbps = mbpsFrom(map["uploadMbps"]),
                jitterMs = (map["jitterMs"] as? Number)?.toDouble() ?: 0.0,
                packetLoss = (map["packetLoss"] as? Number)?.toDouble() ?: 0.0,
                isOffline = map["isOffline"] as? Boolean ?: false,
            )
        }

        private fun mbpsFrom(value: Any?): Double {
            val number = (value as? Number)?.toDouble() ?: return Double.POSITIVE_INFINITY
            return if (number < 0) Double.POSITIVE_INFINITY else number
        }
    }
}
